### ISSUE TITLE
Accuracy Issue BERT Text-Classifciation

### ISSUE THREAD

**Issue by philschmid** [aws/aws-neuron-sdk#303]
Hey, 

I used the [tutorial](https://github.com/aws/aws-neuron-sdk/blob/master/src/examples/pytorch/bert_tutorial/tutorial_pretrained_bert.ipynb) to compile the `bert-base-cased-finetuned-mrpc` as shown in the tutorial. 
I ran some benchmarks afterward against the glue `mrpc` task to validate the performance of the `traced` model compared to the `default` model. 
The results: 
```bash
{'model_type': 'pytorch', 'accuracy': 0.852, 'f1': 0.8987688098495212, 'total_time': '255426.6ms', 'average_time': '255.4266ms'}

{'model_type': 'neuron', 'accuracy': 0.685, 'f1': 0.8130563798219584, 'total_time': '3023.8999999999996ms', 'average_time': '3.0239ms'}
```
The Accuracy went down by 16,7% to 68,5% and the F1 score went down by 8.6% to 81%. 

The model was traced without any additional configuration.
```python
# Run torch.neuron.trace to generate a TorchScript that is optimized by AWS Neuron
model_neuron = torch.neuron.trace(model, example_inputs_paraphrase)
```

I am going to test a different model and dataset next week, but the loss in performance seems to be too high to use in production.

Have you also run some benchmarks with it? Could you share the results of it?

**Comment by philschmid** [comment]
To convert the model i used a docker container, with this `dockerfile`
```dockerfile
FROM continuumio/miniconda3

ENV PYTHON_VERSION=3.7

RUN conda install -y python=$PYTHON_VERSION

RUN  conda config --env --add channels https://conda.repos.neuron.amazonaws.com \
 && conda install torch-neuron=1.7

RUN pip install transformers "datasets>=1.5.0" sklearn neuron-cc[tensorflow]

WORKDIR /home/neuron

RUN conda install -c conda-forge jupyterlab jupyterlab_widgets ipywidgets

CMD ["jupyter", "lab", "--port=8888", "--no-browser", "--ip=0.0.0.0", "--allow-root"]
```
The `neuron-cc` version installed is the following 
```bash
# neuron-cc -V
Neuron Compiler version 1.3.7.0+fc8e67ab4

HWM version 1.3.1.0-0
NEFF version 1.0
TVM version 1.3.1.0+0
NumPy version 1.19.2
MXNet not available
TF not available
ONNX not available
```

**Comment by aws-joshim** [comment]
@philschmid thanks for sharing the details. We re-ran our accuracy benchmarks with the tutorial https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/pytorch/bert_tutorial/tutorial_pretrained_bert.html on DLAMI v48. We see accuracy in the range of 0.845 with the data set listed on the tutorial. In order to recreate the accuracy gap that you see, can you share the details of the specific dataset that you used in your benchmark and the associated accuracy benchmark scripts?

**Comment by philschmid** [comment]
Hey @aws-joshim ,

I used this notebook https://github.com/philschmid/transformers-inference-experiments/blob/main/aws-neuron-sdk/simple_mrpc_example.ipynb

**Comment by aws-joshim** [comment]
Hi Phil, we re-ran the neuron benchmarks with the dataset and scripts @ https://github.com/philschmid/transformers-inference-experiments/blob/main/aws-neuron-sdk/simple_mrpc_example.ipynb,  we see accuracy in the 0.85 range. We suspect a setup difference between your environment and the tutorial that leads to the accuracy degradation. I suggest we run the comparison offline and report back here once we have a conclusion.

**Comment by philschmid** [comment]
Hey @aws-joshim, I tried it again but still get `0.685` accuracy. Here are the steps to reproduce my results. 

Installation instruction I used are from here: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-intro/install-pytorch.html

1. Start ec2 instance (default Amazon Linux2 AMI) 
eu-west-1: `Amazon Linux 2 AMI (HVM), SSD Volume Type - ami-02b4e72b17337d6c1 (64-bit x86)`
instance-type: `inf1.xlarge`
2.  **Configuring Linux for repository updates**

```bash
sudo tee /etc/yum.repos.d/neuron.repo > /dev/null <<EOF
[neuron]
name=Neuron YUM Repository
baseurl=https://yum.repos.neuron.amazonaws.com
enabled=1
metadata_expire=0
EOF

sudo rpm --import https://yum.repos.neuron.amazonaws.com/GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB
sudo yum install kernel-devel-**$(**uname -r**)** kernel-headers-**$(**uname -r**)**
sudo yum install aws-neuron-dkms
sudo yum install aws-neuron-runtime-base
sudo yum install aws-neuron-runtime
sudo yum install aws-neuron-tools
```

1. update libraries

```bash
# Amazon Linux
sudo yum update
sudo yum install -y gcc-c++
```

1. install neuron

```
pip3 install --upgrade --extra-index-url https://pip.repos.neuron.amazonaws.com  neuron-cc[tensorflow] torch-neuron
```

1. install git

```bash
sudo yum install git -y
```

1. install transformers

```bash
pip3 install transformers==4.6.0 datasets "jupyter notebook" sklearn
```

1. tunnel jupyter notebook or run script

```bash
ssh -i "infinity-test.pem" ec2-user@ec2-34-254-204-183.eu-west-1.compute.amazonaws.com -L 8888:127.0.0.1:8888
```

1. script to evaluate
    1. convert model

        ```python
        import tensorflow  # to workaround a protobuf version conflict issue
        import torch
        import torch.neuron
        from transformers import AutoTokenizer, AutoModelForSequenceClassification, AutoConfig
        import transformers

        # Build tokenizer and model
        tokenizer = AutoTokenizer.from_pretrained("bert-base-cased-finetuned-mrpc")
        model = AutoModelForSequenceClassification.from_pretrained("bert-base-cased-finetuned-mrpc", return_dict=False)

        # Setup some example inputs
        sequence_0 = "The company HuggingFace is based in New York City"
        sequence_1 = "Apples are especially bad for your health"
        sequence_2 = "HuggingFace's headquarters are situated in Manhattan"

        max_length=128
        paraphrase = tokenizer.encode_plus(sequence_0, sequence_2, max_length=max_length, padding='max_length', truncation=True, return_tensors="pt")
        not_paraphrase = tokenizer.encode_plus(sequence_0, sequence_1, max_length=max_length, padding='max_length', truncation=True, return_tensors="pt")

        # Run the original PyTorch model on compilation exaple
        paraphrase_classification_logits = model(**paraphrase)[0]

        # Convert example inputs to a format that is compatible with TorchScript tracing
        example_inputs_paraphrase = paraphrase['input_ids'], paraphrase['attention_mask'], paraphrase['token_type_ids']
        example_inputs_not_paraphrase = not_paraphrase['input_ids'], not_paraphrase['attention_mask'], not_paraphrase['token_type_ids']

        # Run torch.neuron.trace to generate a TorchScript that is optimized by AWS Neuron
        model_neuron = torch.neuron.trace(model, example_inputs_paraphrase)

        # Verify the TorchScript works on both example inputs
        paraphrase_classification_logits_neuron = model_neuron(*example_inputs_paraphrase)
        not_paraphrase_classification_logits_neuron = model_neuron(*example_inputs_not_paraphrase)

        # Save the TorchScript for later use
        model_neuron.save('bert_neuron.pt')
        ```

    2. run script

        ```python
        from datasets import load_dataset,load_metric
        import tensorflow  # to workaround a protobuf version conflict issue
        import torch
        import torch.neuron
        from transformers import AutoTokenizer, AutoModelForSequenceClassification
        import time

        task = "mrpc"
        split="validation"
        all_datasets = load_dataset("glue", task)
        metric = load_metric("glue", task)
        dataset= all_datasets[split]

        tokenizer = AutoTokenizer.from_pretrained("bert-base-cased-finetuned-mrpc")
        max_length=128
        padding='max_length'

        def preprocess_function(examples):
            # Tokenize the texts
            texts = (examples['sentence1'], examples['sentence2'])
            result = tokenizer(*texts, padding=padding, max_length=max_length, truncation=True,return_tensors="pt")
            result["labels"] = examples["label"]
            return result

        def do_test(raw_dataset,model,model_type):
            processed_dataset = raw_dataset.map(preprocess_function)
            processed_dataset = processed_dataset.select(range(1000))
            model_start = time.perf_counter()
        #     model_type = 'neuron' if isinstance(model, torch.jit.ScriptModule) else 'torch'
            with torch.no_grad():
                for step, batch in enumerate(processed_dataset):
                    input_ids = torch.tensor(batch['input_ids'])
                    attention_mask = torch.tensor(batch['attention_mask'])
                    token_type_ids = torch.tensor(batch['token_type_ids'])
                    outputs = model(*[input_ids,attention_mask,token_type_ids])
                    predictions = outputs[0][0].argmax().item()
                    metric.add_batch(predictions=[predictions],references=[batch["labels"]])
                
            eval_metric = metric.compute()
            model_stop = time.perf_counter()
            total_time = round(model_stop - model_start,4)*1000
            average_time =  round(total_time/len(processed_dataset),4)
            return {'model_type':model_type,**eval_metric,'total_time':f"{total_time}ms",'average_time':f"{average_time}ms"}   

        model_neuron = torch.jit.load('bert_neuron.pt')
        model = AutoModelForSequenceClassification.from_pretrained("bert-base-cased-finetuned-mrpc", return_dict=False)

        model_res=do_test(dataset, model,'pytorch')
        model_neuron_res = do_test(dataset, model_neuron,'neuron')

        print(model_res)
        print(model_neuron_res)
        ```

    Result

    ```python
    {'model_type': 'pytorch', 'accuracy': 0.852, 'f1': 0.8987688098495212, 'total_time': '116155.7ms', 'average_time': '116.1557ms'}
    {'model_type': 'neuron', 'accuracy': 0.685, 'f1': 0.8130563798219584, 'total_time': '1015.0999999999999ms', 'average_time': '1.0151ms'}
    ```

**Comment by aws-joshim** [comment]
@philschmid thanks for sharing the details of your reproduction steps. We will recreate the same steps on our side and update you on our findings

**Comment by philschmid** [comment]
Reason `aws-neuron-runtime-base` ,`aws-neuron-runtime`,`aws-neuron-tools` didn't get installed properly. Needed to add a `-y` or install each package separatly.

new results
```bash
{'model_type': 'pytorch', 'accuracy': 0.852, 'f1': 0.8987688098495212, 'total_time': '188560.4ms', 'average_time': '188.5604ms'}
{'model_type': 'neuron', 'accuracy': 0.852, 'f1': 0.8987688098495212, 'total_time': '9292.5ms', 'average_time': '9.2925ms'}
```
